File: modules/sql/app/controllers/sql/SqlController.java
package controllers.sql;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import controllers.core.ExerciseController;
import model.CorrectionException;
import model.Levenshtein;
import model.SqlSolution;
import model.SqlSolutionKey;
import model.SqlUser;
import model.StringConsts;
import model.exercise.SqlExercise;
import model.exercise.SqlExerciseType;
import model.exercise.SqlSample;
import model.exercise.SqlScenario;
import model.logging.ExerciseStartEvent;
import model.querycorrectors.SqlResult;
import model.sql.SqlQueryResult;
import model.user.User;
import play.Logger;
import play.data.FormFactory;
import play.db.Database;
import play.db.NamedDatabase;
import play.mvc.Result;
import play.twirl.api.Html;

public class SqlController extends ExerciseController<SqlExercise, SqlResult> {

  private static final String STANDARD_SQL = "";

  private Database sqlSelect;
  private Database sqlOther;

  @Inject
  public SqlController(FormFactory theFactory, @NamedDatabase("sqlselectroot") Database theSqlSelect,
      @NamedDatabase("sqlotherroot") Database theSqlOther) {
    super(theFactory, "sql", SqlExercise.finder);
    sqlSelect = theSqlSelect;
    sqlOther = theSqlOther;
  }

  public static User getUser() {
    User user = ExerciseController.getUser();

    if(SqlUser.finder.byId(user.name) == null)
      // Make sure there is a corresponding entrance in other db...
      new SqlUser(user.name).save();

    return user;
  }

  private static SqlSample findBestFittingSample(String userStatement, List<SqlSample> samples) {
    return samples.parallelStream().min((samp1, samp2) -> {
      int dist1 = Levenshtein.distance(samp1.sample, userStatement);
      int dist2 = Levenshtein.distance(samp2.sample, userStatement);
      return dist1 - dist2;
    }).orElse(samples.get(0));
  }

  private static List<String> readExistingTables(Connection connection) {
    try(ResultSet existingTables = connection.createStatement().executeQuery(StringConsts.SHOW_ALL_TABLES)) {
      final List<String> tableNames = new LinkedList<>();
      while(existingTables.next())
        tableNames.add(existingTables.getString(1));
      return tableNames;
    } catch (SQLException e) {
      Logger.error("There was an error reading the table names from a database", e);
      return Collections.emptyList();
    }
  }

  private static SqlQueryResult readTableContent(Connection connection, String tableName) {
    try(PreparedStatement selectStatement = connection.prepareStatement(StringConsts.SELECT_ALL_DUMMY + tableName)) { // NOSONAR
      final ResultSet tableResult = selectStatement.executeQuery();
      return new SqlQueryResult(tableResult, tableName);
    } catch (SQLException e) {
      Logger.error("There has been an error reading the data from the table " + tableName, e);
      return null;
    }
  }

  private static List<SqlQueryResult> readTablesInDatabase(Database db, String databaseName) {
    try(Connection connection = db.getConnection()) {
      connection.setCatalog(databaseName);
      return readExistingTables(connection).stream().map(tableName -> readTableContent(connection, tableName))
          .collect(Collectors.toList());
    } catch (SQLException e) {
      Logger.error("Es gab einen Fehler beim Auslesen der Tabellen!", e);
      return Collections.emptyList();
    }
  }

  private static void saveSolution(String userName, String learnerSolution, int id) {
    SqlSolutionKey key = new SqlSolutionKey(userName, id);

    SqlSolution solution = SqlSolution.finder.byId(key);
    if(solution == null)
      solution = new SqlSolution(key);

    solution.sol = learnerSolution;
    solution.save();
  }

  public Result exercise(int id) {
    User user = getUser();
    
    SqlExercise exercise = SqlExercise.finder.byId(id);
    
    if(exercise == null)
      return redirect(controllers.sql.routes.SqlController.index());
    
    List<SqlQueryResult> tables = readTablesInDatabase(sqlSelect,
        "" /* exercise.scenario.shortName */);
    
    SqlSolution oldSol = SqlSolution.finder.byId(new SqlSolutionKey(user.name, id));
    String oldOrDefSol = oldSol == null ? STANDARD_SQL : oldSol.sol;
    
    log(user, new ExerciseStartEvent(request(), id));
    
    return ok(views.html.sqlExercise.render(user, exercise, oldOrDefSol, tables));

  }

  public Result filteredScenario(int id, String exType, int start) {
    final SqlScenario scenario = SqlScenario.finder.byId(id);

    if(scenario == null)
      return redirect(controllers.sql.routes.SqlController.index());

    return ok(views.html.sqlscenario.render(getUser(), scenario, SqlExerciseType.valueOf(exType), start));
  }

  public Result index() {
    return ok(views.html.sqlIndex.render(getUser(), SqlScenario.finder.all()));
  }

  public Result scenarioes() {
    return ok(views.html.scenarioes.render(getUser(), SqlScenario.finder.all()));
  }

  private Database getDBForExType(SqlExerciseType exerciseType) {
    return exerciseType == SqlExerciseType.SELECT ? sqlSelect : sqlOther;
  }

  @Override
  protected List<SqlResult> correct(String learnerSolution, SqlExercise exercise, User user)
      throws CorrectionException {
    String learnerSol = factory.form().bindFromRequest().get(StringConsts.FORM_VALUE);

    if(learnerSol == null || learnerSol.isEmpty())
      throw new CorrectionException(learnerSol, StringConsts.EMPTY_SOLUTION);

    saveSolution(user.name, learnerSol, exercise.getId());

    SqlSample sample = findBestFittingSample(learnerSol, exercise.samples);

    return Arrays
        .asList(exercise.getCorrector().correct(getDBForExType(exercise.exerciseType), learnerSol, sample, exercise));
  }

  @Override
  protected Html renderResult(List<SqlResult> correctionResult) {
    // TODO Auto-generated method stub
    return null;
  }
}

File: modules/sql/app/controllers/sql/SqlController.scala
package controllers.sql

import java.sql.Connection
import javax.inject.Inject

import controllers.excontrollers.IdExController
import controllers.sql.SqlController._
import model.CommonUtils.cleanly
import model.StringConsts.{SELECT_ALL_DUMMY, SHOW_ALL_TABLES}
import model._
import model.exercise.{SqlExercise, SqlExerciseType, SqlSample, SqlScenario}
import model.querycorrectors.{QueryCorrector, SqlResult}
import model.result.{CompleteResult, EvaluationResult}
import model.sql.SqlQueryResult
import model.user.User
import play.api.data.Form
import play.api.mvc.{AnyContent, ControllerComponents, Request, Session}
import play.db.{Database, NamedDatabase}
import play.twirl.api.Html

import scala.collection.JavaConverters.asScalaBufferConverter
import scala.collection.mutable.ListBuffer
import scala.util.{Failure, Success, Try}

class SqlController @Inject()(cc: ControllerComponents, @NamedDatabase("sqlselectroot") sqlSelect: Database, @NamedDatabase("sqlotherroot") sqlOther: Database)
  extends IdExController[SqlExercise, EvaluationResult](cc, SqlExercise.finder, SqlToolObject) {

  override type SolType = StringSolution

  override def solForm: Form[StringSolution] = ???

  override def getUser(implicit request: Request[AnyContent]): User = {
    val user = super.getUser

    // Make sure there is a corresponding entrance in other db...
    Option(SqlUser.finder.byId(user.name)) match {
      case None => new SqlUser(user.name).save()
      case Some(_) => Unit
    }

    user
  }

  def filteredScenario(id: Int, exType: String, site: Int) = Action { implicit request =>
    Option(SqlScenario.finder.byId(id)) match {
      case None => Redirect(controllers.sql.routes.SqlController.index())
      case Some(scenario) =>
        if (site <= 0)
          Redirect(controllers.sql.routes.SqlController.filteredScenario(id, exType))
        else {
          val start = SqlScenario.STEP * (site - 1)

          val allByType = scenario.getExercisesByType(SqlExerciseType.valueOf(exType))
          val exercises = allByType.subList(start, Math.min(start + SqlScenario.STEP, allByType.size))

          Ok(views.html.sqlScenario.render(getUser, exercises.asScala.toList, scenario, SqlExerciseType.valueOf(exType), site))
        }
    }
  }

  def index = Action { implicit request => Ok(views.html.sqlIndex.render(getUser, SqlScenario.finder.all.asScala.toList)) }

  def scenarioes = Action { implicit request => Ok(views.html.scenarioes.render(getUser, SqlScenario.finder.all.asScala.toList)) }

  def getDBForExType(exerciseType: SqlExerciseType): Database = if (exerciseType == SqlExerciseType.SELECT) sqlSelect else sqlOther

  override def correctEx(sol: StringSolution, exercise: SqlExercise, user: User): Try[SqlResult] = Try({
    val learnerSolution = sol.learnerSolution
    saveSolution(user.name, learnerSolution, exercise.id)

    val sample = findBestFittingSample(learnerSolution, exercise.samples.asScala.toList)

    val corrector = QueryCorrector.getCorrector(exercise.exerciseType)
    val db = getDBForExType(exercise.exerciseType)

    corrector.correct(db, learnerSolution, sample, exercise)
  })

  override def renderExercise(user: User, exercise: SqlExercise): Html = {
    val tables = readTablesInDatabase(sqlSelect, exercise.scenario.shortName)

    val oldOrDefSol = Option(SqlSolution.finder.byId(new SqlSolutionKey(user.name, exercise.id))) match {
      case None => ""
      case Some(oldSol) => oldSol.sol
    }

    views.html.sqlExercise.render(user, exercise, oldOrDefSol, tables)
  }

  override def renderExesListRest: Html = ??? // FIXME: implement


  override def renderResult(correctionResult: CompleteResult[EvaluationResult]): Html = ??? //FIXME: implement

}

object SqlController {
  def findBestFittingSample(userSt: String, samples: List[SqlSample]): SqlSample = {
    samples.reduceLeft((samp1, samp2) =>
      if (Levenshtein.distance(samp1.sample, userSt) < Levenshtein.distance(samp2.sample, userSt))
        samp1 else samp2)
  }

  def readExistingTables(connection: Connection): List[String] = cleanly(connection.createStatement.executeQuery(SHOW_ALL_TABLES))(_.close)(existingTables => {
    var tableNames = ListBuffer.empty[String]
    while (existingTables.next) {
      tableNames += existingTables.getString(1)
    }
    tableNames.toList
  }) match {
    case Success(list) => list
    case Failure(_) => List.empty
  }

  def readTableContent(connection: Connection, tableName: String): SqlQueryResult =
    cleanly(connection.prepareStatement(SELECT_ALL_DUMMY + tableName))(_.close)(result => new SqlQueryResult(result.executeQuery, tableName)) match {
      case Success(result) => result
      case Failure(_) => null
    }

  def readTablesInDatabase(db: Database, databaseName: String): List[SqlQueryResult] = cleanly(db.getConnection)(_.close)(connection => {
    connection.setCatalog(databaseName)
    readExistingTables(connection).map(readTableContent(connection, _))
  }) match {
    case Success(queryResult) => queryResult
    case Failure(_) => List.empty
  }

  def saveSolution(userName: String, learnerSolution: String, id: Int) {
    val key = new SqlSolutionKey(userName, id)
    var solution = Option(SqlSolution.finder.byId(key)).getOrElse(new SqlSolution(key))

    solution.sol = learnerSolution
    solution.save()
  }
}
